modules/opentelemetry/src/main/scala/zio/opentelemetry/tracing/ContextPropagation.scala

```diff
@@ -0,0 +1,70 @@
+package zio.opentelemetry.tracing
+
+import io.grpc.Context
+import io.opentelemetry.context.propagation.HttpTextFormat
+import io.opentelemetry.context.propagation.HttpTextFormat.{ Getter, Setter }
+import io.opentelemetry.trace.{ Span, TracingContextUtils }
+import zio.{ RIO, Semaphore, UIO }
+
+private[opentelemetry] object ContextPropagation {
+
+  private def getter[C](read: PropagationFormat.Reader[C]): Getter[C] =
+    (carrier: C, key: PropagationFormat.Key) => read(carrier, key).orNull
+
+  private def setter[C](write: PropagationFormat.Writer[C]): Setter[C] =
+    (carrier: C, key: PropagationFormat.Key, value: PropagationFormat.Value) => write(carrier, key, value)
+
+  //  The OpenTelemetry Java API forces us to deal with `Context` when extracting and injecting Spans.
+  //  The context is normally retrieved through `Context.current()`.
+  //  This creates and uses a ThreadLocalStorage to store the span and poses a complication,
+  //  as in the zio landscape we are using FiberRefs to do that instead.
+  //
+  //  A solution is brought by the following observation:
+  //  When extracting a span from a carrier, the context is merely a 'medium' used to retrieve the span.
+  //  This means that OpenTelemetry's function 'extract':
+  //  1) does the actual extraction of the span from the carrier without needing any context
+  //  2) puts the extracted span into the context that was passed to it, and returns that context to the caller
+  //
+  //  When injecting a span into a carrier, the context also serves only as a 'medium' to retrieve the span.
+  //  This means that OpenTelemetry's function 'inject':
+  //  1) retrieves the span from the context that was passed to it.
+  //  2) then does the actual injection of the retrieved span into the carrier, without needing any context
+  //
+  //  Thus we can use a 'dummy' context to play the role of medium: `Context.Root`, which doesn't have any thread local storage.
+  //  However, given there is only a single instance of `Context.Root`, we prevent any synchronization issues by
+  //  permitting access only to a single fiber at a time.
+  //
+  // See https://github.com/open-telemetry/opentelemetry-java/issues/575, and
+  // https://github.com/open-telemetry/opentelemetry-java/issues/1104
+
+  private val semaphore: UIO[Semaphore] = Semaphore.make(1)
+
+  /**
+   * Extract and returns the span from carrier `C`.
+   */
+  def extractSpan[C](
+    httpTextFormat: HttpTextFormat,
+    carrier: C,
+    reader: PropagationFormat.Reader[C]
+  ): UIO[Span] =
+    semaphore.flatMap(_.withPermit {
+      UIO(httpTextFormat.extract(Context.ROOT, carrier, getter(reader))).map(TracingContextUtils.getSpan)
```

Comment:
I think this needs to be uninterruptible. Otherwise, even with the semaphore, something might concurrently set a different span on the thread this gets scheduled on, or the getSpan might get executed on a different thread. That would even work without the Semaphore.

Reply:
Fixed with https://github.com/zio/zio-telemetry/pull/119/commits/85286fb80663bdd87a20d4556812c37d657da7ec

Reply:
Fixed with https://github.com/zio/zio-telemetry/commit/85286fb80663bdd87a20d4556812c37d657da7ec

I didn't know about this feature yet, but looks like it might solve the problem. Ideally we would have a test that dictates having this feature. But given that this problem is likely to be resolved in the next release of Opentelemetry-java (https://github.com/open-telemetry/opentelemetry-java/issues/575), we will not need this soon anymore anyway